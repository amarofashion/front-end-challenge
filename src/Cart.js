import React from 'react';
import Cart from './containers/Cart';
import Header from './containers/Header';
import './index.css';

const App = () => {
    return (
        <div className="mdl-layout mdl-js-layout mdl-color--grey-100">
            <Header />
            <div className="cart-content">
                <Cart />
            </div>
        </div>
    );
}

export default App;
