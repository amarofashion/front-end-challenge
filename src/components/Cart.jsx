import React from 'react';
import { Component } from 'react';
import PropTypes from 'prop-types';

import CartStore from '../stores/Cart.js';

export default class Cart extends Component {
	static propTypes = {
		get: PropTypes.func,
		allProducts: PropTypes.array,
		remove: PropTypes.func
	}

	static defaultProps = {
		get: () => {},
		allProducts: [],
		remove: () => {}
	}

	constructor(props) {
		super(props);
		this.state = {
			products: [],
			showList: false
		};

		this.onChange = this.onChange.bind(this);
		this.toggleList = this.toggleList.bind(this);
		this.remove = this.remove.bind(this);
	}

	componentDidMount() {
		CartStore.listen(this.onChange);
		this.props.get();
	}

	componentWillUnmount() {
		CartStore.unlisten(this.onChange);
	}

	onChange(store) {
		this.setState({
			products: store.products
		});
	}

	toggleList() {
		this.setState({
			showList: !this.state.showList
		});
	}

	remove(evt) {
		this.props.remove(this.state.products[evt.target.dataset.idPosition])
	}

	renderProductsList() {
		const products = this.props.allProducts;
		
		if(!products.length || !this.state.products.length)
			return null;

		return this.state.products.map((id, i) => {
			return <li key={ i } className="product"><a className="link" href={ `#p${id}` }>{ products[id].name }</a><span className="remove" data-id-position={ i } onClick={ this.remove }>X</span></li>
		})
	}

	render() {
		const styleHeight = this.state.showList ? {height: this.state.products.length * 25} : {};
		return (<div className="cart-box">
			<span className="mycart" onClick={ this.toggleList }>Meu carrinho ({ this.state.products.length })</span>
			<ul className={ `products ${this.state.showList ? 'open' : ''}` } style={ styleHeight }>{ this.renderProductsList() }</ul>
		</div>);
	}
}