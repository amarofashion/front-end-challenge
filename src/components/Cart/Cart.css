.cart {
  display: inline-block;
  height: 100%;
  background: #FFF;
  width: 100%;
}

.cart ul {
  border-top: 1px solid #ccc;
  border-bottom: 1px solid #ccc;
  display: inline-block;
  width: 100%;
  padding: 0;
  overflow-y: auto;
  overflow-x: hidden;
  height: 87%;
  position: fixed;
}

.cart h3 {
  padding: 0 20px;
}

.cart ul li {
  list-style: none;
  padding: 10px;
  border-bottom: 1px solid #ccc;
}

.cart ul li:last-child {
  border: none;
}
.cart-item {
  margin-bottom: .3em;
  position: relative;
  display: inline-block;
  width: 100%;
}

.cart-item__qtd,
.cart-item__price,
.cart-item__name {
  display: block;
}

.cart__total {
  font-weight: bold;
  font-size: 1.5em;
  line-height: 1.1em; 
  text-align: right;
  padding: 10px;
  background-color: #000;
  color: #FFF;
  position: fixed;
  width: 100%;
  bottom: 0;
  left: 0;
}

.cart__total p {
  padding: 10px;
  margin: 0 20px 0 0;
  font-size: 20px;
}

.btn-close {
  font-size: 12px;
  background: none;
  border: none;
  position: absolute;
  top: 0;
  right: 0;
}

.cart-item .cart-thumbnail {
  width: 15%;
  float: left;
} 

.empty-cart {
  padding: 40px;
  text-align: center;
}

.empty-cart h3 {
  padding: 0;
}

.bt-buy-more {
  margin: 0 auto;
  border: solid 1px #000;
  padding: 10px;
  display: inline-flex;
  color: #000;
  background-color: #FFF;
}

.bt-buy-more:hover {
  text-decoration: none;
  color: #000;
}

.cart-details {
  float: left;
  width: 70%;
  padding: 10px;
}

.description {
  padding: 15px;
  min-height: 70px;;
}

.item-description {
  width: 100%;
}

.cart-container {
  width: 100%;
}