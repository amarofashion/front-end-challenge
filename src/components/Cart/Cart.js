import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import CartItem from '../../containers/CartItem';

const Cart = ({ items, total, currency, quantity }) => {
    return (
        <div className="cart container">
            <h3>Sacola ({ quantity })</h3>
            <div className="cart-container">
                {items.length > 0 && (
                    <ul>
                        {items.map(item => (
                            <li key={item.id}>
                                <CartItem {...item} /> 
                            </li>
                        ))}
                    </ul>
                )}
                {items.length === 0 && (
                    <div className="empty-cart">
                        <h3>SUA SACOLA EST√Å VAZIA :(</h3>
                        <Link className="bt-buy-more" title="Comprar mais" to="/">Comprar mais</Link>
                    </div>
                )}
            </div>
            <div className="cart__total">
                <p>Total: {currency} {total}</p>
            </div>
        </div>
    );
}

Cart.propTypes = {
    items: PropTypes.array,
    total: PropTypes.string,
    currency: PropTypes.string,
    quantity: PropTypes.number
}

export default Cart;
