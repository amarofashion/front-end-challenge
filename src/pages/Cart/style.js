import styled from 'styled-components';
import breakpoint from 'styled-components-breakpoint';

import colors from '../../theme/colors';

export const StyledWrapper = styled.div`
  .list,
  .checkout {
    display: inline-block;
    vertical-align: top;
  }

  .list {
    width: 100%;
    display: inline-flex;
    flex-wrap: wrap;

    ${breakpoint('tablet')`
      width: 70%;
      display: inline-flex;
    `}

    & > div {
      flex-grow: 1;
      width: 100%;
      margin: 0 0 20px;

      ${breakpoint('tablet')`
        width: calc(50% - 20px);
        margin: 0 10px 20px;
      `}

      ${breakpoint('desktop')`
        width: calc(33% - 20px);
      `}
    }

    & > div:last-child,
    & > div:nth-last-child(2) {
      flex-grow: 0;
    }
  }

  .checkout {
    width: 100%;
    display: block;
    border: 1px solid #ccc;
    padding: 20px;

    ${breakpoint('tablet')`
      width: 30%;
      display: inline-block;
    `}

    .columnLeft,
    .columnRight {
      width: 50%;
      display: inline-block;
      text-align: center;

      p {
        margin-bottom: 10px;
      }

      .total {
        font-size: 18px;
        font-weight: bold;
      }
    }

    .submit {
      button {
        padding: 10px;
        width: 100%;
        display: block;
        background: ${colors.primary};
        color: ${colors.light};
        font-size: 22px;
        font-weight: bold;
        margin: 30px 0 0;
        letter-spacing: 5px;

        &:hover {
          cursor: pointer;
        }
      }
    }
  }
`;

export const StyledSize = styled.div``;
