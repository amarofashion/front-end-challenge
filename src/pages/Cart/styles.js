import styled from 'styled-components';
import { lighten, darken } from 'polished';

export const Container = styled.div`
  footer {
    margin-top: 30px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;

    button {
      background: #fff;
      color: #000;
      border: 0;
      border-radius: 4px;
      padding: 12px 20px;
      font-weight: bold;  
      text-transform: uppercase;
      transition: background 0.2s;

      &:hover {
        background: ${darken(0.2, '#fff')}
      }
    }
    @media (min-width: 769px) 
    {
      flex-direction: row;
    }
  }
`;

export const Grid = styled.div`
  display: flex;
  justify-content: space-between;
  flex-direction: column;
`

export const ProductTable = styled(Grid)`
  .body-item {
    text-align: center;
    padding: 10px;
    background: #fff;
    border-radius: 4px;
    margin-bottom: 10px;
  }

  .product-description {
    display: flex;
    align-items: flex-start;
    justify-content: space-between;
  }

  .product-subtotal {
    padding-top: 10px;
    margin-top: 10px;
    align-items: center;
    flex-direction: row;
    border-top: 1px solid #eee;
    
    p {
      font-weight: bold;
      font-size: 16px;
    }
  }

  img {
    max-width: 110px;
  }

  strong {
    color: #333;
    display: block;
  }

  p {
    margin-top: 5px;
    font-size: 14px;
  }

  .product-quantity {
    display: flex;
    align-items: center;

    input {
      border: 1px solid #ddd;
      border-radius: 4px;
      color: #666;
      padding: 6px;
      width: 50px;
      text-align: center;
    }
  }

  button {
    background: none;
    border: 0;
    padding: 6px;
  }

  .md {
    display: none;
  }

  @media (min-width: 769px) 
  {
    .product-subtotal {
      display: none;
    }

    .md {
      display: flex;
    }

    .product-description {
      align-items: center;
    }

    p {
      font-size: 16px;
    }
  }
`;

export const Total = styled.div`
  display: flex;
  justify-content: center;
  align-items: baseline;
  padding-top: 10px;

  span {
    color: #999;
    font-weight: bold;
  }

  strong {
    font-size: 28px;
    margin-left: 5px;
    color: white;
  }

  @media (min-width: 769px) 
  {
    padding-top: 0px;
  }
`;
