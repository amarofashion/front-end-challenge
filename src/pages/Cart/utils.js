import { translateToNumber } from '../../utils/number';

export function getPrices(list) {
  const result = {
    subTotal: 0,
    totalDiscount: 0,
    total: 0,
  };

  list.forEach(({ onSale, actualPrice, regularPrice, quantity }) => {
    const actualPriceValue = translateToNumber(actualPrice) * quantity;
    const regularPriceValue = translateToNumber(regularPrice) * quantity;

    if (onSale) {
      result.totalDiscount += regularPriceValue - actualPriceValue;
      result.total += actualPriceValue;
    } else {
      result.total += regularPriceValue;
    }

    result.subTotal += regularPriceValue;
  });

  return result;
}
