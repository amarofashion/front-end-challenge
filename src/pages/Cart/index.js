import React from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  MdAddCircleOutline,
  MdRemoveCircleOutline,
  MdDelete
} from "react-icons/md";
import * as CartActions from "../../store/modules/Cart/actions";

import { formatPrice } from "../../util/format";
import { Container, Grid, ProductTable, Total } from "./styles";
import Logo from '../../assets/Amaro_logo.png'

export default function Cart() {

  const cart = useSelector(state => state.cart.map(product => {
    return ({
      ...product,
      subtotal: formatPrice(parseFloat(product.actual_price.replace('R$', '').replace(',', '.')) * product.amount)
    })
  }));
  const total = useSelector(state => formatPrice(
    state.cart.reduce((total, product) => {
      return total + parseFloat(product.actual_price.replace('R$', '').replace(',', '.')) * product.amount;
    }, 0)
  ));

  const dispatch = useDispatch();

  function increment(product, idx) {
    dispatch(CartActions.updateAmountRequest(product.amount + 1, idx));
  }

  function decrement(product, idx) {
    dispatch(CartActions.updateAmountRequest(product.amount - 1, idx));
  }

  return (
    <Container>
      <ProductTable>
        <Grid>
          {cart.map((product, idx) => {
            return (
              <Grid className="body-item" key={product.code_color}>
                <div className="product-description">
                  <img src={product.image || Logo} alt={product.name} />
                  <div>
                    <p>{product.name}</p>
                    <p>Unit√°rio: {product.actual_price}</p>
                  </div>
                  <div className="product-quantity md">
                    <button type="button" onClick={() => decrement(product, idx)}>
                      <MdRemoveCircleOutline size={20} color="#000" />
                    </button>
                    <input readOnly type="number" value={product.amount} />
                    <button type="button" onClick={() => increment(product, idx)}>
                      <MdAddCircleOutline size={20} color="#000" />
                    </button>
                  </div>
                  <p className="md">{product.subtotal}</p>
                  <button
                    type="button"
                    onClick={() => dispatch(CartActions.removeFromCartRequest(product.code_color))}
                  >
                    <MdDelete size={20} color="#000" />
                  </button>
                </div>
                <Grid className="product-subtotal">
                  <div className="product-quantity">
                    <button type="button" onClick={() => decrement(product, idx)}>
                      <MdRemoveCircleOutline size={20} color="#000" />
                    </button>
                    <input readOnly type="number" value={product.amount} />
                    <button type="button" onClick={() => increment(product, idx)}>
                      <MdAddCircleOutline size={20} color="#000" />
                    </button>
                  </div>
                  <p>{product.subtotal}</p>
                </Grid>
              </Grid>
            );
          })}
        </Grid>
      </ProductTable>
      <footer>
        <button type="button">FINALIZAR PEDIDO</button>
        <Total>
          <span>TOTAL</span>
          <strong>{total}</strong>
        </Total>
      </footer>
    </Container>
  );
}
