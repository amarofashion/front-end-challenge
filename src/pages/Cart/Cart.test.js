import React from 'react';
import { mount } from 'enzyme';

import Cart from '.';
import { WithStore, WithRouter } from '../../utils/testing';

describe('<Cart />', () => {
  let wrapper;

  describe('with default props', () => {
    beforeAll(() => {
      wrapper = mount(
        <WithStore>
          <WithRouter>
            <Cart />
          </WithRouter>
        </WithStore>
      );
    });

    it('shoud match snapshot with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
