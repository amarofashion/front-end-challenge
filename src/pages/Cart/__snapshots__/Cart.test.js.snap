// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> with default props shoud match snapshot with default props 1`] = `
<WithStore>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <WithRouter>
      <BrowserRouter>
        <Router
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
        >
          <Cart>
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": false,
                      "lastClassName": "ierQoz",
                      "rules": Array [
                        "
  .list,
  .checkout {
    display: inline-block;
    vertical-align: top;
  }

  .list {
    width: 100%;
    display: inline-flex;
    flex-wrap: wrap;

    ",
                        [Function],
                        "

    & > div {
      flex-grow: 1;
      width: 100%;
      margin: 0 0 20px;

      ",
                        [Function],
                        "

      ",
                        [Function],
                        "
    }

    & > div:last-child,
    & > div:nth-last-child(2) {
      flex-grow: 0;
    }
  }

  .checkout {
    width: 100%;
    display: block;
    border: 1px solid #ccc;
    padding: 20px;

    ",
                        [Function],
                        "

    .columnLeft,
    .columnRight {
      width: 50%;
      display: inline-block;
      text-align: center;

      p {
        margin-bottom: 10px;
      }

      .total {
        font-size: 18px;
        font-weight: bold;
      }
    }

    .submit {
      button {
        padding: 10px;
        width: 100%;
        display: block;
        background: ",
                        "#E56A54",
                        ";
        color: ",
                        "#fff",
                        ";
        font-size: 22px;
        font-weight: bold;
        margin: 30px 0 0;
        letter-spacing: 5px;

        &:hover {
          cursor: pointer;
        }
      }
    }
  }
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "div",
                    "toString": [Function],
                    "usesTheme": false,
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-bwzfXH ierQoz"
                >
                  <div>
                    <p>
                      Pesquise 
                      <Link
                        to="/products"
                      >
                        <LinkAnchor
                          href="/products"
                          navigate={[Function]}
                        >
                          <a
                            href="/products"
                            onClick={[Function]}
                          >
                            aqui
                          </a>
                        </LinkAnchor>
                      </Link>
                       para adicionar mais produtos!
                    </p>
                  </div>
                  <div
                    className="list"
                    data-cypress="cart-list"
                  />
                </div>
              </StyledComponent>
            </styled.div>
          </Cart>
        </Router>
      </BrowserRouter>
    </WithRouter>
  </Provider>
</WithStore>
`;
